File: src/main/java/org/tud/oas/fca/Simple2SFCA.java
package org.tud.oas.fca;

import org.tud.oas.population.Population;
import org.tud.oas.population.PopulationAttributes;
import org.tud.oas.routing.IsochroneCollection;
import org.tud.oas.routing.Isochrone;
import org.tud.oas.routing.IRoutingProvider;

import org.locationtech.jts.geom.Point;
import org.locationtech.jts.geom.Polygon;
import org.locationtech.jts.geom.Geometry;
import org.locationtech.jts.geom.Location;
import org.locationtech.jts.geom.Coordinate;
import org.locationtech.jts.geom.Envelope;
import org.locationtech.jts.algorithm.locate.SimplePointInAreaLocator;

import java.util.List;
import java.util.ArrayList;

public class Simple2SFCA {
    public static float[] calc2SFCA(Population population, Double[][] facilities, List<Double> ranges, List<Double> range_factors, IRoutingProvider provider) throws Exception {
        float[] population_weights = new float[population.getPointCount()];
        float[] facility_weights = new float[facilities.length];
        for (PopulationAttributes attr : population.attributes) {
            population_weights[attr.getIndex()] = attr.getPopulationCount();
        }

        float max_range = ranges.get(ranges.size()-1).floatValue();

        ArrayList<FacilityReference>[] inverted_mapping = new ArrayList[population.getPointCount()];
        Double[][] locations = new Double[1][2];
        for (int f=0; f<facilities.length; f++) {
            locations[0][0] = facilities[f][0];
            locations[0][1] = facilities[f][1];
            IsochroneCollection isochrones = provider.requestIsochrones(locations, ranges).get(0);

            float weight = 0;
            for (int i=0; i< isochrones.getIsochronesCount(); i++) {
                Isochrone isochrone = isochrones.getIsochrone(i);
                double range = isochrone.getValue();
                double range_factor = range_factors.get(ranges.indexOf(range));
                Geometry iso;
                Geometry outer = isochrone.getGeometry();
                if (i==0) {
                    iso = outer;
                }
                else {
                    Geometry inner = isochrones.getIsochrone(i-1).getGeometry();
                    iso = outer.difference(inner);
                }
                Envelope env = iso.getEnvelopeInternal();
                List<Integer> points = population.getPointsInEnvelop(env);
                for (Integer index : points) {
                    Coordinate p = population.getPoint(index);
                    int location = SimplePointInAreaLocator.locate(p, iso);
                    if (location == Location.INTERIOR) {
                        weight += population_weights[index] * range_factor;

                        if (inverted_mapping[index] == null) {
                            inverted_mapping[index] = new ArrayList<FacilityReference>(4);
                        }
                        inverted_mapping[index].add(new FacilityReference(f, (float)range));
                    }
                }
            }
            if (weight == 0) {
                facility_weights[f] = 0;
            }
            else {
                facility_weights[f] = 1/weight;
            }
        }

        for (int p=0; p<population.getPointCount(); p++) {
            List<FacilityReference> refs = inverted_mapping[p];
            float weight = -1;
            if (refs == null) {
                population_weights[p] = weight;
            }
            else {
                weight = 0;
                for (FacilityReference ref : refs) {
                    double range_factor = range_factors.get(ranges.indexOf((double)ref.range));
                    weight += (float)(facility_weights[ref.index] * range_factor);
                }
                population_weights[p] = weight;
            }
        }

        return population_weights;
    }
}

class FacilityReference {
    public int index;
    public float range;

    public FacilityReference(int index, float range) {
        this.index = index;
        this.range = range;
    }
}

File: src/fca/Simple2SFCA.cs
using System;
using System.Collections.Generic;
using NetTopologySuite.Geometries;
using NetTopologySuite.Algorithm.Locate;
using System.Threading.Tasks;
using DVAN.Routing;
using DVAN.Population;
using System.Threading.Tasks.Dataflow;

namespace DVAN.FCA
{
    public class Simple2SFCA
    {
        public static async Task<Dictionary<int, float>> calc2SFCA(PopulationView population, double[][] facilities, List<double> ranges, List<double> range_factors, IRoutingProvider provider)
        {
            var population_weights = new Dictionary<int, float>();
            float[] facility_weights = new float[facilities.Length];

            var inverted_mapping = new Dictionary<int, List<FacilityReference>>();

            var collection = provider.requestIsochronesStream(facilities, ranges);
            for (int f = 0; f < facilities.Length; f++) {
                var isochrones = await collection.ReceiveAsync();
                if (isochrones == null) {
                    continue;
                }

                var visited = new HashSet<int>(10000);
                float weight = 0;
                for (int i = 0; i < isochrones.getIsochronesCount(); i++) {
                    var isochrone = isochrones.getIsochrone(i);
                    double range = isochrone.getValue();
                    double range_factor = range_factors[ranges.IndexOf(range)];

                    Geometry iso = isochrone.getGeometry();
                    Envelope env = iso.EnvelopeInternal;

                    List<int> points = population.getPointsInEnvelop(env);
                    foreach (int index in points) {
                        if (visited.Contains(index)) {
                            continue;
                        }
                        Coordinate p = population.getCoordinate(index);
                        var location = SimplePointInAreaLocator.Locate(p, iso);
                        if (location == Location.Interior) {
                            int population_count = population.getPopulationCount(index);
                            weight += population_count * (float)range_factor;

                            if (!inverted_mapping.ContainsKey(index)) {
                                inverted_mapping[index] = new List<FacilityReference>(4);
                            }
                            inverted_mapping[index].Add(new FacilityReference(f, (float)range));
                            visited.Add(index);
                        }
                    }
                }
                if (weight == 0) {
                    facility_weights[f] = 0;
                }
                else {
                    facility_weights[f] = 1 / weight;
                }
            }

            foreach (int index in inverted_mapping.Keys) {
                List<FacilityReference> refs = inverted_mapping[index];
                if (refs == null) {
                    continue;
                }
                else {
                    float weight = 0;
                    foreach (FacilityReference fref in refs) {
                        double range_factor = range_factors[ranges.IndexOf(fref.range)];
                        weight += (float)(facility_weights[fref.index] * range_factor);
                    }
                    population_weights[index] = weight;
                }
            }

            return population_weights;
        }
    }

    class FacilityReference
    {
        public int index;
        public float range;

        public FacilityReference(int index, float range)
        {
            this.index = index;
            this.range = range;
        }
    }
}